### PR TITLE
feat(euler): create euler integration

### PR DESCRIPTION
**Pull request by conwayconstar**
## Description

Added Zapper integration to Euler Finance.
- Added D Token fetcher
- Added E Token fetcher
- Added P Token fetcher
- Added Balance fetcher for all tokens
- Added TVL fecther

## Checklist

- [x] I have followed the [Contributing Guidelines](../CONTRIBUTING.md)
- [x] (optional) As a contributor, my Ethereum address/ENS is: 0x2F1d02cc9331d8eaDC376A25c283fc5818F95426
- [x] (optional) As a contributor, my Twitter handle is: ConwayConstar

### CODE REVIEW

File: src/apps/euler/ethereum/euler.d-token.token-fetcher.ts [hunk 1]
```diff
@@ -0,0 +1,114 @@
...
+            {
+              type: ContractType.BASE_TOKEN as const,
+              address: market.id,
+              symbol: market.symbol,
+              name: market.name,
+              price: Number(market.twap) / 10 ** 18,
+              network,
+              decimals: Number(market.decimals),
+            },
+          ],
```
Comment (immasandwich) [inline src/apps/euler/ethereum/euler.d-token.token-fetcher.ts:98]: You'll want to get this token from the base tokens:

```
const baseTokens = await this.appToolkit.getBaseTokenPrices(network);

// ...
const underlyingToken = tokens.find(t => t.address === market.id.toLowerCase());

// ...
tokens: [underlyingToken]
```

File: src/apps/euler/ethereum/euler.d-token.token-fetcher.ts [hunk 2]
```diff
@@ -0,0 +1,114 @@
...
+          dataProps: {
+            name: `Euler D token ${market.name}`,
+            interestRate: Number(market.interestRate) / 10 ** 18,
+            borrowAPY: Number(market.borrowAPY) / 10 ** 18,
+            supplyAPY: Number(market.borrowAPY) / 10 ** 18,
+          },
+          displayProps: { label: market.name, images: [''] },
```
Comment (immasandwich) [inline src/apps/euler/ethereum/euler.d-token.token-fetcher.ts:105]: `label` should be `Euler D Token ${market.name}`, and `images` should be `getImagesFromToken(underlyingToken!)`.

You can also add a `secondaryLabel` and set that to the price of the D token like `buildDollarDisplayItem(Number(market.twap) / 10 ** 18)`

File: src/apps/euler/ethereum/euler.d-token.token-fetcher.ts (hunk 1 above)
Comment (conwayconstar) [reply src/apps/euler/ethereum/euler.d-token.token-fetcher.ts:98]: twap is what we use for the price from uniswap within our API, will use the helper from appToolKit for this instead!

File: src/apps/euler/ethereum/euler.d-token.token-fetcher.ts [hunk 3]
```diff
@@ -0,0 +1,113 @@
...
+
+        if (totalSupply.isZero() || !underlyingToken) return null;
+
+        return {
+          address: market.dTokenAddress,
+          symbol: `D${market.symbol}`,
+          name: `Euler D token ${market.name}`,
+          type: ContractType.APP_TOKEN as const,
+          supply: Number(market.totalSupply) / 10 ** Number(market.decimals),
+          pricePerShare: underlyingToken.price,
```
[the remark this answers is not in the document]
Comment (immasandwich) [reply src/apps/euler/ethereum/euler.d-token.token-fetcher.ts:89]: i.e.: `pricePerShare === price / underlyingToken.price` (its a "ratio")

[the remark this answers is not in the document]
Comment (immasandwich) [reply src/apps/euler/ethereum/euler.d-token.token-fetcher.ts:89]: (Once you update the pricePerShare, I think we're good! Will merge it right after)

**Pull request merged**